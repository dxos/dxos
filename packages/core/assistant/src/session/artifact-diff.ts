//
// Copyright 2025 DXOS.org
//

import { Context } from 'effect';

import { type ObjectVersion } from '@dxos/echo-db';
import { type ObjectId } from '@dxos/echo-schema';

/**
 * Resolves artifact ids to their versions.
 * Used to give the model a sense of the changes to the artifacts made by users during the conversation.
 * The artifacts versions are pinned in the history, and whenever the artifact changes in-between assistant's steps,
 * a diff is inserted into the conversation.
 *
 * Can be optionally provided to the session run call.
 */
// TODO(burdon): Move.
// TODO(dmaretskyi): Convert to Context.Reference
export class ArtifactDiffResolver extends Context.Tag('@dxos/assistant/ArtifactDiffResolver')<
  ArtifactDiffResolver,
  ArtifactDiffResolver.Service
>() {}

export namespace ArtifactDiffResolver {
  export type Service = {
    resolve: (artifacts: { id: ObjectId; lastVersion: ObjectVersion }[]) => Promise<
      Map<
        ObjectId,
        {
          version: ObjectVersion;
          diff?: string;
        }
      >
    >;
  };
}
