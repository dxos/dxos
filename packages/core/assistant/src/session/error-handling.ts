//
// Copyright 2025 DXOS.org
//

import { AiError } from '@effect/ai';
import { HttpClientError } from '@effect/platform';
import * as Effect from 'effect/Effect';
import * as Schema from 'effect/Schema';

import { AiModelError } from '../errors';

const AnthropicErrorResponse = Schema.Struct({
  type: Schema.Literal('error'),
  error: Schema.Struct({
    type: Schema.String,
    message: Schema.String,
  }),
});

// TODO(dmaretskyi): Needs rework
export const mapAiError = (err: AiError.AiError): Effect.Effect<AiError.AiError> =>
  Effect.gen(function* () {
    const cause = err.cause;
    if (HttpClientError.isHttpClientError(cause) && cause.reason === 'StatusCode') {
      const body = yield* cause.response.json.pipe(
        Effect.flatMap(Schema.decodeUnknown(AnthropicErrorResponse, { exact: false })),
      );

      const parsedCause = new AiModelError({
        message: body.error.message,
        context: { model: 'anthropic', type: body.error.type },
      });
      return AiError.UnknownError.make({
        description: body.error.message,
        module: err.module,
        method: err.method,
        cause: parsedCause,
      });
    }
    return err;
  }).pipe(Effect.catchAll(() => Effect.succeed(err)));
