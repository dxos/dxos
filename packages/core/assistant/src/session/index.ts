//
// Copyright 2025 DXOS.org
//

export * from './artifact-diff';
export * from './observer';
export * from './session';
export * from './toolkit';
export * from './version-pin';
