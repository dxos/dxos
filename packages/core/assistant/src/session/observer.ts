//
// Copyright 2025 DXOS.org
//

import { type AiResponse } from '@effect/ai';
import { Effect, Function } from 'effect';

import { type ConsolePrinter } from '@dxos/ai';
import { type ContentBlock, type DataType } from '@dxos/schema';

/**
 * Live observer of the generation process.
 */
export interface GenerationObserver {
  /**
   * Unparsed content block parts from the model.
   */
  onPart: (part: AiResponse.Part) => Effect.Effect<void>;

  /**
   * Parsed content blocks from the model.
   * NOTE: Use block.pending to determine if the block was completed.
   * For each block this will be called 0..n times with a pending block and then once with the final state of the block.
   *
   * Example:
   *  1. { pending: true, text: "Hello"}
   *  2. { pending: true, text: "Hello, I am a"}
   *  3. { pending: false, text: "Hello, I am a helpful assistant!"}
   */
  onBlock: (block: ContentBlock.Any) => Effect.Effect<void>;

  /**
   * Complete messages fired during the session, both from the model and from the user.
   * This message will contain all message blocks emitted through the `onBlock` callback.
   */
  onMessage: (message: DataType.Message) => Effect.Effect<void>;
}

export const GenerationObserver = Object.freeze({
  make: ({
    onPart = Function.constant(Effect.void),
    onBlock = Function.constant(Effect.void),
    onMessage = Function.constant(Effect.void),
  }: Partial<GenerationObserver> = {}): GenerationObserver => ({
    onPart,
    onBlock,
    onMessage,
  }),

  noop: () => GenerationObserver.make(),

  /**
   * Debug printer to be used in unit-tests and browser devtools.
   */
  fromPrinter: (printer: ConsolePrinter) =>
    GenerationObserver.make({
      onBlock: (block) =>
        Effect.sync(() => {
          if (block.pending) {
            return; // Only prints full blocks (better for unit-tests and browser devtools).
          }
          printer.printContentBlock(block);
        }),
      onMessage: (message) =>
        Effect.sync(() => {
          if (message.sender.role === 'assistant') {
            return; // Skip assistant messages since they are printed in the `onBlock` callback.
          }
          printer.printMessage(message);
        }),
    }),
});
