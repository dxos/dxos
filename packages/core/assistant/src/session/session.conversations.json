{"conversations":[{"parameters":{"model":"@anthropic/claude-3-5-sonnet-20241022","tools":[]},"history":{"content":[{"role":"user","content":"Hello world!","options":{}},{"role":"assistant","content":"Hi","options":{}},{"role":"assistant","content":"!","options":{}},{"role":"assistant","content":" How can I help","options":{}},{"role":"assistant","content":" you today?","options":{}}]}},{"parameters":{"model":"@anthropic/claude-3-5-sonnet-20241022","tools":[{"name":"Calculator","description":"Basic calculator tool","inputSchema":{"type":"object","required":["input"],"properties":{"input":{"type":"string","description":"The calculation to perform."}},"additionalProperties":false}}]},"history":{"content":[{"role":"user","content":"What is 10 + 20?","options":{}},{"role":"assistant","content":"I'll","options":{}},{"role":"assistant","content":" help","options":{}},{"role":"assistant","content":" you calculate","options":{}},{"role":"assistant","content":" ","options":{}},{"role":"assistant","content":"10 + 20","options":{}},{"role":"assistant","content":" using the","options":{}},{"role":"assistant","content":" Calculator tool","options":{}},{"role":"assistant","content":".","options":{}},{"role":"assistant","content":[{"type":"tool-call","id":"toolu_01HKtEU7wz4KSpqcCNRnwUXj","name":"Calculator","params":{"input":"10 + 20"},"providerExecuted":false,"options":{}}],"options":{}}]}},{"parameters":{"model":"@anthropic/claude-3-5-sonnet-20241022","tools":[{"name":"Calculator","description":"Basic calculator tool","inputSchema":{"type":"object","required":["input"],"properties":{"input":{"type":"string","description":"The calculation to perform."}},"additionalProperties":false}}]},"history":{"content":[{"role":"user","content":"What is 10 + 20?","options":{}},{"role":"assistant","content":[{"type":"text","text":"I'll help you calculate 10 + 20 using the Calculator tool.","options":{}},{"type":"tool-call","id":"toolu_01HKtEU7wz4KSpqcCNRnwUXj","name":"Calculator","params":{"input":"10 + 20"},"providerExecuted":false,"options":{}}],"options":{}},{"role":"tool","content":[{"type":"tool-result","id":"toolu_01HKtEU7wz4KSpqcCNRnwUXj","name":"Calculator","result":{"result":30},"options":{}}],"options":{}},{"role":"assistant","content":"The result of ","options":{}},{"role":"assistant","content":"10 + 20","options":{}},{"role":"assistant","content":" is 30.","options":{}}]}},{"parameters":{"model":"@anthropic/claude-3-5-sonnet-20241022","tools":[{"name":"Calculator","description":"Basic calculator tool","inputSchema":{"type":"object","required":["input"],"properties":{"input":{"type":"string","description":"The calculation to perform."}},"additionalProperties":false}}]},"history":{"content":[{"role":"user","content":"What is 10 + 30?","options":{}},{"role":"assistant","content":"I","options":{}},{"role":"assistant","content":"'ll","options":{}},{"role":"assistant","content":" help","options":{}},{"role":"assistant","content":" you calculate","options":{}},{"role":"assistant","content":" ","options":{}},{"role":"assistant","content":"10 + 30","options":{}},{"role":"assistant","content":" using the","options":{}},{"role":"assistant","content":" Calculator","options":{}},{"role":"assistant","content":" tool.","options":{}},{"role":"assistant","content":[{"type":"tool-call","id":"toolu_01U7x664APyAGpzJo6mgRtxm","name":"Calculator","params":{"input":"10 + 30"},"providerExecuted":false,"options":{}}],"options":{}}]}},{"parameters":{"model":"@anthropic/claude-3-5-sonnet-20241022","tools":[{"name":"Calculator","description":"Basic calculator tool","inputSchema":{"type":"object","required":["input"],"properties":{"input":{"type":"string","description":"The calculation to perform."}},"additionalProperties":false}}]},"history":{"content":[{"role":"user","content":"What is 10 + 30?","options":{}},{"role":"assistant","content":[{"type":"text","text":"I'll help you calculate 10 + 30 using the Calculator tool.","options":{}},{"type":"tool-call","id":"toolu_01U7x664APyAGpzJo6mgRtxm","name":"Calculator","params":{"input":"10 + 30"},"providerExecuted":false,"options":{}}],"options":{}},{"role":"tool","content":[{"type":"tool-result","id":"toolu_01U7x664APyAGpzJo6mgRtxm","name":"Calculator","result":{"result":40},"options":{}}],"options":{}},{"role":"assistant","content":"The result of ","options":{}},{"role":"assistant","content":"10 + 30","options":{}},{"role":"assistant","content":" is 40.","options":{}}]}}]}