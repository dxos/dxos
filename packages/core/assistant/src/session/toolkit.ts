//
// Copyright 2025 DXOS.org
//

import { type AiTool, AiToolkit } from '@effect/ai';
import { type Context, Effect } from 'effect';

import { ToolExecutionService, type ToolId, ToolResolverService } from '@dxos/ai';
import { type Blueprint } from '@dxos/blueprints';
import { isTruthy } from '@dxos/util';

export type ToolkitParams<Tools extends AiTool.Any> = {
  toolkit?: AiToolkit.AiToolkit<Tools>;
  toolIds?: ToolId[];
  blueprints?: Blueprint.Blueprint[];
};

/**
 * Build a combined toolkit from the blueprint tools and the provided toolkit.
 */
export const createToolkit = <Tools extends AiTool.Any>({
  toolkit: toolkitParam,
  toolIds = [],
  blueprints = [],
}: ToolkitParams<Tools>) =>
  Effect.gen(function* () {
    const blueprintToolkit = yield* ToolResolverService.resolveToolkit([
      ...blueprints.flatMap(({ tools }) => tools),
      ...toolIds,
    ]);

    const blueprintToolkitHandler: Context.Context<AiTool.ToHandler<AiTool.Any>> = yield* blueprintToolkit.toContext(
      ToolExecutionService.handlersFor(blueprintToolkit),
    );

    const toolkit = AiToolkit.merge(...[toolkitParam, blueprintToolkit].filter(isTruthy));
    return yield* toolkit.pipe(Effect.provide(blueprintToolkitHandler)) as Effect.Effect<
      AiToolkit.ToHandler<any>,
      never,
      AiTool.ToHandler<Tools>
    >;
  });
