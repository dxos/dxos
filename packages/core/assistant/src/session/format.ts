//
// Copyright 2025 DXOS.org
//

import * as Array from 'effect/Array';
import * as Effect from 'effect/Effect';
import * as Function from 'effect/Function';
import * as Option from 'effect/Option';

import { Template } from '@dxos/blueprints';
import { Obj } from '@dxos/echo';
import { type ObjectId } from '@dxos/echo/internal';
import { ObjectVersion } from '@dxos/echo-db';
import { DatabaseService } from '@dxos/functions';
import { log } from '@dxos/log';
import { type ContentBlock, DataType } from '@dxos/schema';
import { trim } from '@dxos/util';

import { AiAssistantError } from '../errors';

import { ArtifactDiffResolver } from './artifact-diff';
import { type AiSessionRunParams } from './session';

/**
 * Formats the system prompt.
 */
// TODO(burdon): Move to AiPreprocessor.
export const formatSystemPrompt = ({
  system,
  blueprints = [],
  objects = [],
}: Pick<AiSessionRunParams<any>, 'system' | 'blueprints' | 'objects'>) =>
  Effect.gen(function* () {
    const blueprintDefs = yield* Function.pipe(
      blueprints,
      Effect.forEach((blueprint) => Effect.succeed(blueprint.instructions)),
      Effect.flatMap(Effect.forEach((template) => DatabaseService.loadOption(template.source))),
      Effect.map(Array.filter(Option.isSome)),
      Effect.map(
        Array.map(
          (template) => trim`
            <blueprint>
              ${Template.process(template.value.content)}
            </blueprint>
          `,
        ),
      ),
      Effect.map((blueprints) =>
        blueprints.length > 0 ? ['## Blueprints Definitions', ...blueprints].join('\n\n') : undefined,
      ),
    );

    const objectDefs = yield* Function.pipe(
      objects,
      Effect.forEach((object) =>
        Effect.succeed(trim`
          <object>
            <dxn>${Obj.getDXN(object)}</dxn>
            <typename>${Obj.getTypename(object)}</typename>
          </object>
        `),
      ),
      Effect.map((objects) => (objects.length > 0 ? ['## Context Objects', ...objects].join('\n\n') : undefined)),
    );

    return yield* Function.pipe(
      Effect.succeed([system, blueprintDefs, objectDefs].filter((def): def is string => def !== undefined)),
      Effect.map((parts) => parts.join('\n\n')),
    );
  });

/**
 * Formats the user prompt.
 */
// TODO(burdon): Move to AiPreprocessor.
// TODO(burdon): Convert util below to `Effect.fn` (to preserve stack info)
export const formatUserPrompt = ({ prompt, history = [] }: Pick<AiSessionRunParams<any>, 'prompt' | 'history'>) =>
  Effect.gen(function* () {
    const blocks: ContentBlock.Any[] = [];

    // TODO(dmaretskyi): Evaluate other approaches as `serviceOption` isn't represented in the type system.
    const artifactDiffResolver = yield* Effect.serviceOption(ArtifactDiffResolver);
    if (Option.isSome(artifactDiffResolver)) {
      const versions = gatherObjectVersions(history);
      const artifactDiff = yield* Effect.tryPromise({
        try: () =>
          artifactDiffResolver.value.resolve(
            [...versions.entries()].map(([id, version]) => ({ id, lastVersion: version })),
          ),
        catch: AiAssistantError.wrap({ message: 'Artifact diff resolution error' }),
      });

      log('version', { artifactDiff, versions });
      for (const [id, { version }] of [...artifactDiff.entries()]) {
        if (ObjectVersion.equals(version, versions.get(id)!)) {
          artifactDiff.delete(id);
          continue;
        }

        blocks.push({ _tag: 'anchor', objectId: id, version });
      }

      if (artifactDiff.size > 0) {
        blocks.push(createArtifactUpdateBlock(artifactDiff));
      }
    }

    return Obj.make(DataType.Message, {
      created: new Date().toISOString(),
      sender: { role: 'user' },
      blocks: [...blocks, { _tag: 'text', text: prompt }],
    });
  });

const gatherObjectVersions = (messages: DataType.Message[]): Map<ObjectId, ObjectVersion> => {
  const artifactIds = new Map<ObjectId, ObjectVersion>();
  for (const message of messages) {
    for (const block of message.blocks) {
      if (block._tag === 'anchor') {
        artifactIds.set(block.objectId, block.version as ObjectVersion);
      }
    }
  }

  return artifactIds;
};

const createArtifactUpdateBlock = (
  artifactDiff: Map<ObjectId, { version: ObjectVersion; diff?: string }>,
): ContentBlock.Any => {
  return {
    _tag: 'text',
    // TODO(dmaretskyi): Does this need to be a special content-block?
    disposition: 'artifact-update',
    text: trim`
      The following artifacts have been updated since the last message:
      ${[...artifactDiff.entries()]
        .map(([id, { diff }]) => `<changed-artifact id="${id}">${diff ? `\n${diff}` : ''}</changed-artifact>`)
        .join('\n')}
    `,
  };
};
