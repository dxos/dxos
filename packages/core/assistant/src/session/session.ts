//
// Copyright 2025 DXOS.org
//

import { type AiError, AiLanguageModel, type AiTool, type AiToolkit } from '@effect/ai';
import { Chunk, Effect, type Schema, Stream } from 'effect';

import {
  type AiInputPreprocessingError,
  AiParser,
  AiPreprocessor,
  type AiToolNotFoundError,
  type ToolExecutionService,
  type ToolResolverService,
  callTool,
  getToolCalls,
} from '@dxos/ai';
import { type Blueprint } from '@dxos/blueprints';
import { todo } from '@dxos/debug';
import { Obj } from '@dxos/echo';
import { TracingService } from '@dxos/functions';
import { log } from '@dxos/log';
import { DataType } from '@dxos/schema';

import { type AiAssistantError } from '../errors';

import { mapAiError } from './error-handling';
import { formatSystemPrompt, formatUserPrompt } from './format';
import { GenerationObserver } from './observer';

export type AiSessionRunError = AiError.AiError | AiInputPreprocessingError | AiToolNotFoundError | AiAssistantError;

export type AiSessionRunRequirements =
  | AiLanguageModel.AiLanguageModel
  | ToolExecutionService
  | ToolResolverService
  | TracingService;

export type AiSessionRunParams<Tools extends AiTool.Any> = {
  prompt: string;
  system?: string;
  history?: DataType.Message[];
  objects?: Obj.Any[];
  blueprints?: Blueprint.Blueprint[];
  toolkit?: AiToolkit.ToHandler<Tools>;
  observer?: GenerationObserver;
};

export type AiSessionOptions = {};

/**
 * Contains message history, tools, current context.
 * Current context means the state of the app, time of day, and other contextual information.
 * It makes requests to the model, its a state machine.
 * It keeps track of the current goal.
 * It manages the context window.
 * Tracks the success criteria of reaching the goal, exposing metrics (stretch).
 * Could be run locally in the app or remotely.
 * Could be personal or shared.
 */
export class AiSession {
  static run: <Tools extends AiTool.Any>(
    params: AiSessionRunParams<Tools>,
  ) => Effect.Effect<
    DataType.Message[],
    AiAssistantError | AiInputPreprocessingError | AiToolNotFoundError | AiError.AiError,
    | AiLanguageModel.AiLanguageModel
    | ToolResolverService
    | ToolExecutionService
    | TracingService
    | AiTool.ToHandler<Tools>
  > = <Tools extends AiTool.Any>(params: AiSessionRunParams<Tools>) => new AiSession().run(params);
  run = <Tools extends AiTool.Any>({
    prompt,
    system: systemTemplate,
    history = [],
    objects = [],
    blueprints = [],
    toolkit,
    observer = GenerationObserver.noop(),
  }: AiSessionRunParams<Tools>): Effect.Effect<DataType.Message[], AiSessionRunError, AiSessionRunRequirements> =>
    Effect.gen(this, function* () {
      const now = Date.now();
      let toolCount = 0;

      // Reset.
      this._history = [...history];
      this._pending = [];

      // Generate system prompt.
      // TODO(budon): Dynamically resolve template variables here.
      const system = yield* formatSystemPrompt({ system: systemTemplate, blueprints, objects });

      const pending = this._pending;
      const submitMessage = Effect.fnUntraced(function* (message: DataType.Message) {
        pending.push(message);
        yield* observer.onMessage(message);
        yield* TracingService.emitConverationMessage(message);
        return message;
      });

      const promptMessage = yield* submitMessage(yield* formatUserPrompt({ prompt, history }));

      //
      // Tool call loop.
      //
      do {
        log.info('request', {
          prompt: promptMessage,
          system: { snippet: createSnippet(system), length: system.length },
          pending: this._pending.length,
          history: this._history.length,
          objects: objects?.length ?? 0,
        });

        //
        // Make the request.
        //
        const prompt = yield* AiPreprocessor.preprocessAiInput([...this._history, ...this._pending]);
        const blocks = yield* AiLanguageModel.streamText({
          prompt,
          system,
          toolkit,
          // TODO(burdon): Check if this bug has been fixed and update deps/patches?
          // TODO(burdon): Despite this flag, the model still calls tools.
          //  Flag is only used in generateText (not streamText); patch and submit bug.
          //  https://github.com/Effect-TS/effect/blob/main/packages/ai/ai/src/AiLanguageModel.ts#L401
          disableToolCallResolution: true,
        }).pipe(
          Stream.catchTag(
            'AiError',
            Effect.fnUntraced(function* (err) {
              return yield* Effect.fail(yield* mapAiError(err));
            }),
          ),
          AiParser.parseResponse({
            onBlock: (block) => observer.onBlock(block),
            onPart: (part) => observer.onPart(part),
          }),
          Stream.runCollect,
          Effect.map(Chunk.toArray),
        );

        // Create response message.
        const response = yield* submitMessage(
          Obj.make(DataType.Message, {
            created: new Date().toISOString(),
            sender: { role: 'assistant' },
            blocks,
          }),
        );

        // Parse response for tool calls.
        const toolCalls = getToolCalls(response);
        if (toolCalls.length === 0) {
          break;
        }
        if (!toolkit) {
          throw new Error('No toolkit provided'); // TODO(burdon): Throw user error?
        }

        // TODO(burdon): Retry backend errors?
        const toolResults = yield* Effect.forEach(toolCalls, (toolCall) => {
          toolCount++;
          return callTool(toolkit, toolCall).pipe(
            Effect.provide(
              TracingService.layerSubframe((context) => ({
                ...context,
                parentMessage: response.id,
                toolCallId: toolCall.toolCallId,
              })),
            ),
          );
        });

        // Add to queue and continue loop.
        yield* submitMessage(
          Obj.make(DataType.Message, {
            created: new Date().toISOString(),
            sender: { role: 'user' },
            blocks: toolResults,
          }),
        );
      } while (true);

      // Summary.
      yield* submitMessage(
        Obj.make(DataType.Message, {
          created: new Date().toISOString(),
          sender: { role: 'assistant' },
          blocks: [
            {
              _tag: 'summary',
              message: 'Success',
              duration: Date.now() - now,
              toolCalls: toolCount,
              // TODO(burdon): Get token count.
            },
          ],
        }),
      );

      log('done', { pending: this._pending.length });
      return this._pending;
    }).pipe(this._semaphore.withPermits(1), Effect.withSpan('AiSession.run'));

  /** Prevents concurrent execution of session. */
  private readonly _semaphore = Effect.runSync(Effect.makeSemaphore(1));

  /**
   * Prior history from queue.
   * NOTE: The conversation should evolve into supporting a git-like graph of messages.
   */
  private _history: DataType.Message[] = [];

  /**
   * Pending messages for this session (incl. the current prompt).
   */

  private _pending: DataType.Message[] = [];

  constructor(private readonly _options: AiSessionOptions = {}) {}

  // TODO(burdon): Implement.
  async runStructured<S extends Schema.Schema.AnyNoContext>(
    _schema: S,
    _options: AiSessionRunParams<AiTool.Any>,
  ): Promise<Schema.Schema.Type<S>> {
    return todo();
    // const parser = structuredOutputParser(schema);
    // const result = await this.run({
    //   ...options,
    //   executableTools: [...(options.executableTools ?? []), parser.tool],
    // });
    // return parser.getResult(result);
  }
}

const createSnippet = (text: string, len = 32) =>
  text.length <= len * 2 ? text : [text.slice(0, len), '...', text.slice(-len)].join('');
