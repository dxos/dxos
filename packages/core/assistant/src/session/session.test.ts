//
// Copyright 2025 DXOS.org
//

import { Tool, Toolkit } from '@effect/ai';
import { describe, it } from '@effect/vitest';
import { Effect, Layer, Schema } from 'effect';

import { AiService, ToolExecutionService, ToolResolverService } from '@dxos/ai';
import { AiServiceTestingPreset } from '@dxos/ai/testing';
import { Obj, Type } from '@dxos/echo';
import { TracingService } from '@dxos/functions';
import { log } from '@dxos/log';

import { AiSession } from './session';

// Define a calendar event artifact schema.
const CalendarEventSchema = Schema.Struct({
  title: Schema.String,
  startTime: Schema.String,
  endTime: Schema.String,
  description: Schema.String,
}).pipe(
  Type.Obj({
    typename: 'example.com/type/CalendarEvent',
    version: '0.1.0',
  }),
);

type CalendarEvent = Schema.Schema.Type<typeof CalendarEventSchema>;

class TestToolkit extends Toolkit.make(
  Tool.make('Calculator', {
    description: 'Basic calculator tool',
    parameters: {
      input: Schema.String.annotations({
        description: 'The calculation to perform.',
      }),
    },
    success: Schema.Struct({
      result: Schema.Number,
    }),
    failure: Schema.Never,
  }),
) {}

// Tool handlers.
const toolkitLayer = TestToolkit.toLayer({
  Calculator: ({ input }) =>
    Effect.gen(function* () {
      const result = (() => {
        // Restrict to basic arithmetic operations for safety.
        const sanitizedInput = input.replace(/[^0-9+\-*/().\s]/g, '');
        log.info('calculate', { sanitizedInput });

        // eslint-disable-next-line @typescript-eslint/no-implied-eval
        return Function(`"use strict"; return (${sanitizedInput})`)();
      })();

      return { result };
    }),
});

describe.runIf(process.env.DX_RUN_SLOW_TESTS)('AiSession', () => {
  it.effect('no tools', () =>
    Effect.gen(function* () {
      const session = new AiSession({ operationModel: 'configured' });
      const response = yield* session.run({
        prompt: 'Hello world!',
        history: [],
      });
      log.info('response', { response });
    }).pipe(
      Effect.provide(
        Layer.mergeAll(
          toolkitLayer,
          AiService.model('@anthropic/claude-3-5-sonnet-20241022').pipe(
            Layer.provideMerge(ToolResolverService.layerEmpty),
            Layer.provideMerge(ToolExecutionService.layerEmpty),
            Layer.provideMerge(AiServiceTestingPreset('direct')),
            Layer.provideMerge(TracingService.layerNoop),
          ),
        ),
      ),
    ),
  );

  it.effect('calculator', () =>
    Effect.gen(function* () {
      const session = new AiSession({ operationModel: 'configured' });
      const toolkit = yield* TestToolkit;
      const response = yield* session.run({
        toolkit,
        prompt: 'What is 10 + 20?',
        history: [],
      });
      log.info('response', { response });
    }).pipe(
      Effect.provide(
        Layer.mergeAll(
          TracingService.layerNoop,
          AiService.model('@anthropic/claude-3-5-sonnet-20241022').pipe(
            Layer.provideMerge(AiServiceTestingPreset('direct')),
          ),
          ToolResolverService.layerEmpty,
          ToolExecutionService.layerEmpty,
          toolkitLayer,
        ),
      ),
    ),
  );

  // TODO(dmaretskyi): Revive test.
  /*
  it.skip('create calendar itinerary', { timeout: 60_000 }, async () => {
    // overrides: { model: 'llama3.1:8b' },
    const session = new AiSession({ operationModel: 'configured' });
    const objects = new Set<string>();

    // Define calendar artifact.
    const calendarArtifact = defineArtifact({
      id: 'artifact:dxos.org/plugin/calendar',
      name: 'Calendar',
      instructions: 'Use this to create and query calendar events.',
      schema: CalendarEventSchema,
      tools: [
        createTool('calendar', {
          name: 'query',
          description: 'Query the calendar for events',
          schema: Schema.Struct({}),
          execute: async () => {
            return ToolResult.Success(CALENDAR_EVENTS);
          },
        }),
      ],
    });

    const tableArtifact = defineArtifact({
      id: 'artifact:dxos.org/plugin/table',
      name: 'Table',
      instructions: 'Use this to create and manage tables. Each table has a unique id.',
      schema: Schema.Struct({}),
      tools: [
        createTool('table', {
          name: 'create',
          description: 'Create a table',
          schema: Schema.Struct({
            data: Schema.Array(Schema.Any).annotations({ description: 'Array of data payloads to add as rows' }),
          }),
          execute: async ({ data }) => {
            log('create table', { data });
            const id = DXN.fromLocalObjectId(ObjectId.random()).toString();
            objects.add(id);
            // TODO(dmaretskyi): consider xml for refs instead of @dxn:echo:@:XXXXX
            return ToolResult.Success(`table @${id}`);
          },
        }),
      ],
    });

    const mapArtifact = defineArtifact({
      id: 'artifact:dxos.org/plugin/map',
      name: 'Map',
      instructions:
        'Use this to create and manage maps. Maps source data from tables. Table id is required to create a map.',
      schema: Schema.Struct({}),
      tools: [
        createTool('map', {
          name: 'create',
          description: 'Create a map',
          schema: Schema.Struct({
            source: ArtifactId.annotations({
              description: 'The table that will be used as the source of the map',
            }),
          }),
          execute: async ({ source }) => {
            // TODO(dmaretskyi): Use effect-schema decode instead of manual parsing.
            const sourceId = ArtifactId.toDXN(source);
            if (!objects.has(sourceId.toString())) {
              return ToolResult.Error(`table id=${source} not found`);
            }
            log('create map', { sourceId });
            const id = DXN.fromLocalObjectId(ObjectId.random()).toString();
            objects.add(id);
            // TODO(dmaretskyi): consider xml for refs instead of @dxn:echo:@:XXXXX
            return ToolResult.Success(`map @${id}`);
          },
        }),
      ],
    });

    const scriptArtifact = defineArtifact({
      id: 'artifact:dxos.org/plugin/script',
      name: 'Script',
      instructions: 'Use this to create and manage scripts',
      schema: Schema.Struct({}),
      tools: [],
    });

    // session.streamEvent.on((event) => {
    //   printStreamEvent(event);
    // });

    const printer = new ConsolePrinter();
    session.message.on((message) => printer.printMessage(message));
    session.userMessage.on((message) => printer.printMessage(message));
    session.block.on((block) => printer.printContentBlock(block));

    // session.update.on((update) => {
    //   log('update', { update });
    // });

    // Test creating an itinerary
    const response = await session.run({
      client: aiClient,
      tools: [],
      artifacts: [calendarArtifact, tableArtifact, mapArtifact, scriptArtifact],
      requiredArtifactIds: [calendarArtifact.id, tableArtifact.id, mapArtifact.id, scriptArtifact.id],
      history: [],
      generationOptions: {
        model: '@anthropic/claude-3-5-haiku-20241022',
      },
      prompt: 'create a table and map for a travel itinerary based on events in my calendar',
      toolResolver: new ToolRegistry([]),
    });

    log('result', {
      objects,
      finalMessage: response.at(-1),
    });
  });
  */
});

// Travel to rome, florence, livorno, siena, madrid for conferences

const _CALENDAR_EVENTS: CalendarEvent[] = [
  Obj.make(CalendarEventSchema, {
    title: 'Exploring Ancient Ruins in Rome',
    startTime: '2024-01-01T10:00:00Z',
    endTime: '2024-01-01T11:00:00Z',
    description: 'Tech conference at the historic Colosseum with networking opportunities',
  }),
  Obj.make(CalendarEventSchema, {
    title: 'Renaissance Tech Summit in Florence',
    startTime: '2024-01-01T11:00:00Z',
    endTime: '2024-01-01T12:00:00Z',
    description: 'Discussing AI innovations surrounded by Renaissance art',
  }),
  Obj.make(CalendarEventSchema, {
    title: 'Travel to Livorno',
    startTime: '2024-01-01T12:00:00Z',
    endTime: '2024-01-01T13:00:00Z',
    description: 'Travel to Livorno',
  }),
  Obj.make(CalendarEventSchema, {
    title: 'Travel to Siena',
    startTime: '2024-01-01T13:00:00Z',
    endTime: '2024-01-01T14:00:00Z',
    description: 'Travel to Siena',
  }),
  Obj.make(CalendarEventSchema, {
    title: 'Travel to Madrid',
    startTime: '2024-01-01T14:00:00Z',
    endTime: '2024-01-01T15:00:00Z',
    description: 'Travel to Madrid',
  }),
];

// const printStreamEvent = (event: GenerationStreamEvent) => {
//   switch (event.type) {
//     case 'message_start': {
//       process.stdout.write(`${event.message.role.toUpperCase()}\n\n`);
//       for (const content of event.message.content) {
//         printContentBlock(content);
//       }
//       break;
//     }
//     case 'content_block_start': {
//       printContentBlock(event.content);
//       break;
//     }
//     case 'content_block_delta': {
//       switch (event.delta.type) {
//         case 'text_delta': {
//           process.stdout.write(event.delta.text);
//           break;
//         }
//         case 'input_json_delta': {
//           process.stdout.write(event.delta.partial_json);
//           break;
//         }
//       }
//       break;
//     }
//     case 'content_block_stop': {
//       process.stdout.write('\n');
//       break;
//     }
//     case 'message_delta': {
//       break;
//     }
//     case 'message_stop': {
//       process.stdout.write('\n\n');
//       break;
//     }
//   }
// };
