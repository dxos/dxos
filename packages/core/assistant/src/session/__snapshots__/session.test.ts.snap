// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AiSession > no tools 1`] = `
[
  {
    "blocks": ReactiveArray [
      {
        "_tag": "text",
        "text": "Hello world!",
      },
    ],
    "created": "Sun, 20 Jul 2025 10:54:08 GMT",
    "id": "01K0KQGFTFV3GBKGEWZW67K8HK",
    "sender": {
      "role": "user",
    },
  },
  {
    "blocks": ReactiveArray [
      {
        "_tag": "text",
        "text": "Hi there! How can I help you today?",
      },
    ],
    "created": "2025-07-20T10:54:10.801Z",
    "id": "01K0KQGHVHJ37D4YG45WG090WV",
    "sender": {
      "role": "assistant",
    },
  },
]
`;
